### PR TITLE
[#7285] Student: submit a team response: the student's team members should also be added to the session's responding students list

### LINKED ISSUE #7285: Student: submit a team response: the student's team members should also be added to the session's responding students list
Currently only students who submit a response on behalf of the team are added to the list. This could create a situation where team members have to resubmit the same response in order to be considered a respondent.

### PR DESCRIPTION
**Pull request by junming403**
Fixes #7285

**Outline of Solution**

Add a case to handle team response

### CODE REVIEW

**Comment by junming403** [comment]
This fix fails when a special case occurs where teamName = email. Is this a case in the live server? I am thinking of changing the teamName validator such that it cannot be email. Suggestions?

**Comment by xpdavid** [comment]
You can write a script to check whether there is any `teamName` containing the symbol `@`  and let @damithc help to run it.

**Comment by damithc** [comment]
> This fix fails when a special case occurs where teamName = email. Is this a case in the live server? I am thinking of changing the teamName validator such that it cannot be email. Suggestions?

Let's not restrict the nature of the team name unless there is no other option. Users will not like it when we don't allow them to use the team name they chose.

**Comment by xpdavid** [comment]
> Let's not restrict the nature of the team name unless there is no other option.

Actually, this should the discussion when we decided to switch to use team name as a giver identifier for a response four years ago (https://github.com/TEAMMATES/teammates/commit/1ca8f0a6db2cb8a3baab3fb8858fb0ee24334136). If we don't restrict the team name (currently there is a restriction that team name should not contain `|` or `%`), there is no way for us to distinguish between a student and a team. I believe such special case can already cause data inconsistency in V6 (of course also in V7) (e.g. if we want to check whether user `xx@email.com` has any response while there is a team response from team `xx@email.com`). Therefore, we should take care of the naming right now to prevent it become future legacy problem.

**Comment by damithc** [comment]
Sure, we can add `@` to the blacklist if there is no other option.

**Comment by junming403** [comment]
I have added a script for checking is any teamName a valid email in live server. Ready for review

File: src/client/java/teammates/client/scripts/ValidateAllTeamNameScript.java [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+
+/**
+ * Script to check if any team name is a email.
+ *
+ * <p>See issue #7285</p>
+ */
+public class ValidateAllTeamNameScript extends DataMigrationEntitiesBaseScript<CourseStudent> {
+
+    private final Pattern emailRegex =
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/ValidateAllTeamNameScript.java]: Should use `REGEX_EMAIL` in `FieldValidator`?

File: src/client/java/teammates/client/scripts/ValidateAllTeamNameScript.java [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+    @Override
+    protected int getCursorInformationPrintCycle() {
+        return 100;
+    }
+
+    @Override
+    protected boolean isMigrationNeeded(Key<CourseStudent> key) {
+        CourseStudent student = ofy().load().key(key).now();
+        String teamName = student.getTeamName();
+        return teamName.equals(student.getEmail()) || validate(teamName);
```
Comment (xpdavid) [inline src/client/java/teammates/client/scripts/ValidateAllTeamNameScript.java]: Do you really need this `teamName.equals(student.getEmail())`?

File: src/client/java/teammates/client/scripts/ValidateAllTeamNameScript.java (hunk 1 above)
Comment (junming403) [reply src/client/java/teammates/client/scripts/ValidateAllTeamNameScript.java]: I tried, but that would not pass lint check...

Comment (xpdavid) [reply src/client/java/teammates/client/scripts/ValidateAllTeamNameScript.java]: Then just copy the regex.

**Comment by junming403** [comment]
updated, ready for review

**Comment by junming403** [comment]
Updated, using StringHelper library for email check, ready for review.

**Comment by junming403** [comment]
Resolved lint error

**Comment by junming403** [comment]
Ready for review

**Comment by xpdavid** [comment]
@damithc Please help to run `ValidateAllTeamNameScript.java` against live server and post the summary in console here. No sensitive information will be exposed as we just want to know the number of entities affected.

P.S. @junming403 The script should be discarded after Prof Damith finishes running. It won't go into `master`.

**Comment by damithc** [comment]
Running now. After running for 1.5 hours, I have the following:
```
Number Of Entity Key Scanned: 10800
Number Of Entity affected: 0
Number Of Entity updated: 0
```

Should it be this slow? Note that we have 500K CourseStudent objects :-p Is it going to take 75 hours 😟

**Comment by damithc** [comment]
36K entities read after 5 hours. Feels like something is not right. Are we reading one entity at a time when we should be doing some sort of batch reading?

**Comment by xpdavid** [comment]
I guess it is because of transaction. We should have turned it off since this is just a script for existence checking not for data migration.

**Comment by damithc** [comment]
Timed out after 42k entities (less than 10%). Should we do something about the script before running again?

**Comment by junming403** [comment]
Changed shouldUseTransaction to False. Ready for review

**Comment by xpdavid** [comment]
@damithc Try pull the script again and fit `getLastPositionOfCursor` with last position recorded?

**Comment by damithc** [comment]
> @damithc Try pull the script again and fit `getLastPositionOfCursor` with last position recorded?

I'm running it now. The rate seems to be 2300 entities per 10 minutes. Will take 35-40 hours to complete. Still too slow for just a read operation?

**Comment by xpdavid** [comment]
> I'm running it now. The rate seems to be 2300 entities per 10 minutes. Will take 35-40 hours to complete. Still too slow for just a read operation?

Yeah. I think there is still room for improvement. Let the script keep running. Meanwhile I'll try to improve the framework.

**Comment by junming403** [comment]
Hi Prof, did the script finish running yet?

**Comment by damithc** [comment]
> Hi Prof, did the script finish running yet?

Still 🏃 
The good news is no problematic team names reported so far.

**Comment by damithc** [comment]
The script finished just now. No problematic team names.

**Comment by junming403** [comment]
Thank you prof, I have removed the script, ready for review.

**Comment by junming403** [comment]
Updated validator for team name, now team name cannot be a valid email. 
Ready for review.

**Comment by junming403** [comment]
Ready for review.

File: src/main/java/teammates/common/util/FieldValidator.java [hunk 3]
```diff
@@ -144,6 +144,8 @@
             + "The value of '${fieldName}' field should be no longer than ${maxLength} characters.";
     public static final String INVALID_NAME_ERROR_MESSAGE =
             ERROR_INFO + " " + HINT_FOR_CORRECT_FORMAT_FOR_INVALID_NAME;
+    public static final String TEAM_NAME_IS_VALID_EMAIL_ERROR_MESSAGE =
```
Comment (xpdavid) [inline src/main/java/teammates/common/util/FieldValidator.java]: The change to this file should be exactly the same as https://github.com/TEAMMATES/teammates/commit/df87c34d9652d2c1f7fb77b10c45fc03b6656cb2?

File: src/main/java/teammates/ui/webapi/action/ConfirmFeedbackSessionSubmissionAction.java [hunk 4]
```diff
@@ -63,9 +64,19 @@ public ActionResult execute() {
...
         case STUDENT_SUBMISSION:
             StudentAttributes studentAttributes = getStudentOfCourseFromRequest(feedbackSession.getCourseId());
-            boolean hasStudentRespondedForSession =
+            boolean hasIndividualRespondedForSession =
                     logic.hasGiverRespondedForSession(studentAttributes.getEmail(), feedbackSessionName, courseId);
-            if (hasStudentRespondedForSession) {
+            boolean hasTeamRespondedForSession =
+                    logic.hasGiverRespondedForSession(studentAttributes.getTeam(), feedbackSessionName, courseId);
+
+            if (hasTeamRespondedForSession) {
```
Comment (xpdavid) [inline src/main/java/teammates/ui/webapi/action/ConfirmFeedbackSessionSubmissionAction.java]: Let's have test for this?

**Comment by junming403** [comment]
Close this issue for now as it has wait until API migration finish.

**Pull request closed without merging**